modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointLocal.java

```diff
@@ -0,0 +1,119 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.transactions;
+
+import java.util.LinkedHashMap;
+import org.apache.ignite.internal.util.tostring.GridToStringInclude;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Transaction savepoint implementation for near local transactions.
+ */
+public class TxSavepointLocal implements TxSavepoint {
+    /** Savepoint ID */
+    @GridToStringInclude
+    private final String name;
+
+    /** Per-transaction read map. */
+    @GridToStringInclude
+    private final Map<IgniteTxKey, IgniteTxEntry> txMapSnapshot;
+
+    /**
+     * @param name Savepoint ID.
+     * @param tx Transaction, which state should be saved.
+     */
+    public TxSavepointLocal(String name, IgniteInternalTx tx) {
+        this.name = name;
+
+        Collection<IgniteTxEntry> stateEntries = tx.txState().allEntries();
+
+        txMapSnapshot = new LinkedHashMap<>(stateEntries.size());
+
+        putCopies(stateEntries, txMapSnapshot);
+    }
+
+    /** {@inheritDoc} */
+    @Override public String getName() {
+        return name;
+    }
+
+    /**
+     * @return Entries stored in savepoint.
+     */
+    public Map<IgniteTxKey, IgniteTxEntry> getTxMapSnapshot() {
+        return txMapSnapshot;
+    }
+
+    /**
+     * @param key Key object.
+     * @return True if savepoint contains entry with specified key.
+     */
+    public boolean containsKey(IgniteTxKey key) {
+        return txMapSnapshot.containsKey(key);
+    }
+
+    /**
+     * Takes collection with IgniteTxEntries and copy it into specified map.
+     *
+     * @param from Takes this.
+     * @param to And put all members here.
+     */
+    private void putCopies(Collection<IgniteTxEntry> from, Map<IgniteTxKey, IgniteTxEntry> to) {
+        for (IgniteTxEntry entry : from) {
+            IgniteTxEntry e = entry.cleanCopy(entry.context());
+
+            e.cached(entry.cached());
+
+            if (entry.locked())
+                e.markLocked();
+
+            to.put(entry.txKey(), e);
+        }
+    }
+
+    /** Equality of savepoints depends on their IDs.
+     *
+     * @param o Another savepoint.
+     * @return True if savepoints have equal names.
+     * */
+    @Override public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        TxSavepointLocal savepoint = (TxSavepointLocal) o;
+
+        return name.equals(savepoint.name);
+    }
+
+    /**
+     * @return Copy of txMap snapshot.
+     */
+    public Map<IgniteTxKey, IgniteTxEntry> extractValues() {
```

Comment:
may be rename to txMapSnapshot()

Reply:
I think it is bad idea, because method returns not snapshot, but it's copy. Also method returns not map, but only values.

Reply:
Map<IgniteTxKey, IgniteTxEntry> not map?

Reply:
Yes, but Map<>.values() is not.